src/BotPlutusInterface/CardanoNode/Effects.hs

```diff
@@ -0,0 +1,176 @@
+{-# LANGUAGE RankNTypes #-}
+
+{- This is ongoing effort on replacing `cardano-cli` calls with `Cardano.Api` queries, see issues
+  https://github.com/mlabs-haskell/bot-plutus-interface/issues/109
+  https://github.com/mlabs-haskell/bot-plutus-interface/issues/101
+  We decided to provide single replacement for `BotPlutusInterface.CardanoCLI.utxosAt`
+  early on to enable inline Datum support from one side and avoid extending
+  `cardano-cli` output parser from the other side.
+  See https://github.com/mlabs-haskell/bot-plutus-interface/issues/145
+-}
+module BotPlutusInterface.CardanoNode.Effects (
+  utxosAt,
+  pparams,
+  minUtxo,
+  handleNodeQuery,
+  runNodeQuery,
+  NodeQuery (..),
+) where
+
+import BotPlutusInterface.CardanoNode.Query (
+  NodeConn,
+  NodeQueryError,
+  QueryConstraint,
+  connectionInfo,
+  queryBabbageEra,
+  queryInCardanoMode,
+  toQueryError,
+ )
+
+import BotPlutusInterface.CardanoAPI (
+  addressInEraToAny,
+  fromCardanoTxOut,
+ )
+
+import BotPlutusInterface.Types (PABConfig)
+import Cardano.Api (LocalNodeConnectInfo (..))
+import Cardano.Api qualified as CApi
+import Cardano.Api.Shelley qualified as CApi.S
+import Cardano.Ledger.Shelley.API.Wallet (
+  CLI (evaluateMinLovelaceOutput),
+ )
+import Control.Lens (folded, to, (^..))
+import Control.Monad.Freer (Eff, Members, interpret, runM, send, type (~>))
+import Control.Monad.Freer.Reader (Reader, ask, runReader)
+import Control.Monad.Trans.Class (lift)
+import Control.Monad.Trans.Either (firstEitherT, hoistEither, newEitherT, runEitherT)
+import Control.Monad.Trans.Except (throwE)
+import Data.Map (Map)
+import Data.Map qualified as Map
+import Data.Set qualified as Set
+import Ledger qualified
+import Ledger.Ada qualified as Ada
+import Ledger.Address (Address)
+import Ledger.Tx (ChainIndexTxOut (..))
+import Ledger.Tx.CardanoAPI qualified as TxApi
+import Ledger.Validation (Coin (Coin))
+import Plutus.V2.Ledger.Tx qualified as V2
+import Prelude
+
+data NodeQuery a where
```

Comment:
For Me: Add docs and Explain how and what `NodeQuery` does.

Reply:
fixed by: https://github.com/mlabs-haskell/bot-plutus-interface/pull/103/commits/bb053efa4a10540446339b7123e87c423a20453c